oauth2_client/client.py

```diff
@@ -29,9 +38,58 @@ class OAuth2Client(OAuth2Session):
         > r = client.get('/api/license/1/detail/')
     """
 
-    def __init__(self, app, token):
-        self.service_host = app.service_host  # used to transform relative URLs to absolute
-        super(OAuth2Client, self).__init__(client_id=app.client_id, token=token)
+    """
+    If token re-fetch happens more than once in this period, raise an exception. Something is wrong,
+    don't spam the provider.
+    """
+    REAUTH_LIMIT_SECONDS = 10
+
+    def __init__(self, token, expired_callback=None):
+        """
+        Create OAuth2Client
+
+        :param token: oauth2_client.model.AccessToken object
+        :param expired_callback: (optional) function accepting one parameter, `Application`. It is
+            called when TokenExpiredError is detected or 403 Forbidden status code received.
+            It should return a valid (not expired) oauth2_client.model.AccessToken object for
+            communication with Resource Owner specified by `Application`.
+            Explanation: The parent class, OAuth2Session already implements automated token refresh,
+            but their solution works only for flows that issue a refresh_token. Both currently
+            supported flows, JWT and Client Credentials don't provide the refresh_token, so we have
+            to handle the refresh ourselves. (Client Credentials MAY issue refresh_token but it
+            doesn't in case of the `oauth2_provider` we use). If the token expiry is detected more
+            than once within `REAUTH_LIMIT_SECONDS` period, `oauthlib.oauth2.TokenExpiredError`
+            is raised (safety fuse approach).
+        """
+        self.app = token.application  # Application this client talks to
+        self.last_token_fetch_time = None  # used for the safety fuse preventing too often re-authorization
```

Comment:
Since the only thing the code fetches is tokens we could probably use `last_fetch_time`, with two fewer syllables to mentally pronounce as you read the code. Once they become habitual these small things can really add up, even though each alone is tiny. Or maybe as a compromise (and much better if more than one thing is being fetched) `last_token_time` ...

Reply:
outdated due to introduction of 3rd party implementation of a circuit breaker pattern